File: sample-videochat-webrtc/sample-videochat-webrtc/ViewControllers/IncomingCallViewController/IncomingCallViewController.m
//
//  IncomingCallViewController.m
//  QBRTCChatSemple
//
//  Created by Andrey Ivanov on 16.12.14.
//  Copyright (c) 2014 QuickBlox Team. All rights reserved.
//

#import "IncomingCallViewController.h"
#import "QBButton.h"
#import "QMSoundManager.h"
#import "QBToolBar.h"
#import "QBButtonsFactory.h"
#import "CornerView.h"
#import "UsersDataSource.h"

@interface IncomingCallViewController () <QBRTCClientDelegate>

@property (weak, nonatomic) IBOutlet UILabel *callStatusLabel;
@property (weak, nonatomic) IBOutlet UITextView *callInfoTextView;
@property (weak, nonatomic) IBOutlet QBToolBar *toolbar;
@property (weak, nonatomic) IBOutlet CornerView *colorMarker;
@property (strong, nonatomic) NSArray *users;

@property (weak, nonatomic) NSTimer *dialignTimer;

@end

@implementation IncomingCallViewController

- (void)dealloc {
    NSLog(@"%@ - %@",  NSStringFromSelector(_cmd), self);
}

- (void)viewDidLoad {
    [super viewDidLoad];
    [QMSoundManager playRingtoneSound];
    
    [QBRTCClient.instance addDelegate:self];
    NSMutableArray *users = [NSMutableArray array];
    
    for (NSNumber *uID in self.session.opponentsIDs) {
        
        QBUUser *user = [self.usersDatasource userWithID:uID.integerValue];
        
        if (!user) {
            user = [QBUUser user];
            user.ID = uID.integerValue;
        }
        
        [users addObject:user];
    }
    self.users = [users copy];
    
    [self confiugreGUI];
    
    self.dialignTimer =
    [NSTimer scheduledTimerWithTimeInterval:[QBRTCConfig dialingTimeInterval]
                                     target:self
                                   selector:@selector(dialing:)
                                   userInfo:nil
                                    repeats:YES];
}

- (void)dialing:(NSTimer *)timer {
    
    [QMSoundManager playRingtoneSound];
}

#pragma mark - Update GUI

- (void)confiugreGUI {
    
    [self defaultToolbarConfiguration];
    [self updateOfferInfo];
    [self updateCallInfo];
    
    self.title = [NSString stringWithFormat:@"Logged in as %@", [QBChat instance].currentUser.fullName] ;

}

- (void)defaultToolbarConfiguration {
    
    self.toolbar.backgroundColor = [UIColor clearColor];
    __weak __typeof(self)weakSelf = self;
    
    [self.toolbar addButton:[QBButtonsFactory circleDecline] action: ^(UIButton *sender) {
        
        [weakSelf cleanUp];
        [weakSelf.delegate incomingCallViewController:weakSelf didRejectSession:weakSelf.session];
    }];
    
    [self.toolbar addButton:[QBButtonsFactory answer] action: ^(UIButton *sender) {
        
        [weakSelf cleanUp];
        [weakSelf.delegate incomingCallViewController:weakSelf didAcceptSession:weakSelf.session];
    }];
    
    
    [self.toolbar updateItems];
}

- (void)updateOfferInfo {
    
    QBUUser *caller = [self.usersDatasource userWithID:self.session.initiatorID.unsignedIntegerValue];
    
    self.colorMarker.bgColor = [UIColor grayColor];
    self.colorMarker.title = caller.fullName;
    self.colorMarker.fontSize = 30.f;
}

- (void)updateCallInfo {
    
    NSMutableArray *info = [NSMutableArray array];
    
    for (QBUUser *user in self.users ) {
        
        [info addObject:[NSString stringWithFormat:@"%@(ID %@)", user.fullName, @(user.ID)]];
    }
    
    self.callInfoTextView.text = [info componentsJoinedByString:@", "];
    self.callInfoTextView.font = [UIFont fontWithName:@"HelveticaNeue-Thin" size:19];
    self.callInfoTextView.textAlignment = NSTextAlignmentCenter;
    
    NSString *text =
    self.session.conferenceType == QBRTCConferenceTypeVideo ? @"Incoming video call" : @"Incoming audio call";
    self.callStatusLabel.text = NSLocalizedString(text, nil);
}

#pragma mark - Actions

- (void)cleanUp {
    
    [self.dialignTimer invalidate];
    self.dialignTimer = nil;
    
    [QBRTCClient.instance removeDelegate:self];
	[QBRTCSoundRouter.instance deinitialize];
    [QMSysPlayer stopAllSounds];
}

- (void)sessionDidClose:(QBRTCSession *)session {
    
    if (self.session == session) {
        [self cleanUp];
    }
}

@end

File: sample-videochat-webrtc-swift/sample-videochat-webrtc-swift/Components/IncomingCallViewController/IncomingCallViewController.swift
//
//  IncomingCallViewController.swift
//  sample-videochat-webrtc-swift
//
//  Created by Vladimir Nybozhinsky on 12/18/18.
//  Copyright © 2018 QuickBlox. All rights reserved.
//

import UIKit
import QuickbloxWebRTC
import Quickblox

protocol IncomingCallViewControllerDelegate: NSObjectProtocol {
  func incomingCallViewController(_ vc: IncomingCallViewController, didAccept session: QBRTCSession)
  func incomingCallViewController(_ vc: IncomingCallViewController, didReject session: QBRTCSession)
}

class IncomingCallViewController: UIViewController, QBRTCClientDelegate {
  
  @IBOutlet private weak var callStatusLabel: UILabel!
  @IBOutlet private weak var callInfoTextView: UITextView!
  @IBOutlet private weak var toolbar: ToolBar!
  @IBOutlet private weak var colorMarker: CornerView!
  private var users: [QBUUser] = []
  private var dialignTimer: Timer?
  
  weak var delegate: IncomingCallViewControllerDelegate?
  var session: QBRTCSession?
  var usersDatasource: UsersDataSource?
  
  deinit {
    debugPrint("deinit \(self)")
  }
  
  override func awakeFromNib() {
    super.awakeFromNib()
    
    QBRTCClient.instance().add(self)
  }
  
  override func viewDidLoad() {
    super.viewDidLoad()
    SoundManager.playRingtoneSound()
    var users: [QBUUser] = []
    if let session = session,
      let usersDatasource = usersDatasource {
      for uID in session.opponentsIDs {
        if let user = usersDatasource.user(withID: uID.uintValue) {
          users.append(user)
        } else {
          let user = QBUUser()
          user.id = uID.uintValue
          users.append(user)
        }
      }
    }
    self.users = users
    confiugreGUI()
    dialignTimer = Timer.scheduledTimer(timeInterval: QBRTCConfig.dialingTimeInterval(),
                                        target: self,
                                        selector: #selector(dialing(_:)),
                                        userInfo: nil,
                                        repeats: true)
  }
  
  @objc func dialing(_ timer: Timer?) {
    SoundManager.playRingtoneSound()
  }
  
  // MARK: - Update GUI
  func confiugreGUI() {
    defaultToolbarConfiguration()
    updateOfferInfo()
    updateCallInfo()
    if let currentUser = QBChat.instance.currentUser {
      title = "Logged in as \(String(describing: currentUser.fullName))"
    }
  }
  
  func defaultToolbarConfiguration() {
    toolbar.backgroundColor = UIColor.clear
    toolbar.add(ButtonsFactory.circleDecline(), action: { [weak self] sender in
      guard let `self` = self else {
        return
      }
      self.cleanUp()
      if let session = self.session {
        self.delegate?.incomingCallViewController(self, didReject: session)
      }
    })
    
    toolbar.add(ButtonsFactory.answer(), action: { [weak self] sender in
      guard let `self` = self else {
        return
      }
      self.cleanUp()
      if let session = self.session {
        self.delegate?.incomingCallViewController(self, didAccept: session)
      }
    })
    toolbar.updateItems()
  }
  
  // MARK: - Actions
  func cleanUp() {
    dialignTimer?.invalidate()
    dialignTimer = nil
    QBRTCClient.instance().remove(self)
    SoundManager.instance.stopAllSounds()
  }
  
  func updateOfferInfo() {
    
    if let usersDatasource = usersDatasource,
      let session = session,
      let caller = usersDatasource.user(withID: session.initiatorID.uintValue) {
      colorMarker.bgColor = UIColor.gray
      colorMarker.title = caller.fullName ?? ""
      colorMarker.fontSize = 30.0
    }
  }
  
  func updateCallInfo() {
    
    var info: [String] = []
    for user in users {
      info.append("\(String(describing: user.fullName))(ID \(user.id))")
    }
    callInfoTextView.text = info.joined(separator: ", ")
    callInfoTextView.font = UIFont(name: "HelveticaNeue-Thin", size: 19)
    callInfoTextView.textAlignment = .center
    let text = session?.conferenceType == QBRTCConferenceType.video ? "Incoming video call" : "Incoming audio call"
    callStatusLabel.text = NSLocalizedString(text, comment: "")
  }
  
  func sessionDidClose(_ session: QBRTCSession) {
    if self.session == session {
      cleanUp()
    }
  }
}
